rfcs/InputUnion.md

```diff
@@ -0,0 +1,119 @@
+RFC: GraphQL Input Union
+----------
+
+## Possible Solutions
+
+### Value-based Discriminator field
+
+These options rely the _value_ of a particular input field to determine the concrete type.
+
+#### Single `__typename` field; value is the `type`
+
+```graphql
+input SummaryMetric {
+  min: Integer
+  max: Integer
+}
+input CounterMetric {
+  count: Integer
+}
+input HistogramMetric {
+  percentiles: [Integer]
+}
+
+inputUnion MetricInputUnion = SummaryMetric | CounterMetric | HistogramMetric
+
+{__typename: "SummaryMetric", min: 123, max: 456}
+{__typename: "CounterMetric", count: 789}
+{__typename: "HistogramMetric", percentiles: [12, 34, 56, 78, 90]}
+```
+
+#### Single user-chosen field; value is the `type`
+
+```graphql
+input SummaryMetric {
+  metricType: <MetricInputUnion>
+  min: Integer
+  max: Integer
+}
+input CounterMetric {
+  metricType: <MetricInputUnion>
+  count: Integer
+}
+input HistogramMetric {
+  metricType: <MetricInputUnion>
+  percentiles: [Integer]
+}
+
+inputUnion MetricInputUnion = SummaryMetric | CounterMetric | HistogramMetric
+
+{metricType: "SummaryMetric", min: 123, max: 456}
+{metricType: "CounterMetric", count: 789}
+{metricType: "HistogramMetric", percentiles: [12, 34, 56, 78, 90]}
+```
+
+#### Single user-chosen field; value is a literal
+
+```graphql
+enum MetricType {
+  SUMMARY
+  COUNTER
+  HISTOGRAM
+}
+input SummaryMetric {
+  metricType: MetricType::SUMMARY
+  min: Integer
+  max: Integer
+}
+input CounterMetric {
+  metricType: MetricType::COUNTER
+  count: Integer
+}
+input HistogramMetric {
+  metricType: MetricType::HISTOGRAM
+  percentiles: [Integer]
+}
+
+inputUnion MetricInputUnion = SummaryMetric | CounterMetric | HistogramMetric
+
+{metricType: SUMMARY, min: 123, max: 456}
+{metricType: COUNTER, count: 789}
+{metricType: HISTOGRAM, percentiles: [12, 34, 56, 78, 90]}
+```
+
+### Structural Discrimination
+
+These options rely on the _structure_ of the input to determine the concrete type.
+
+#### Unique structure
```

Comment:
This solution only considers the set of required fields. There's a further generalization that also considers the type of each field and can discriminate fields that have the same name but non-overlapping value domains. For example you could have the following if you wanted to define a filtering API:

```graphql
input StringFilter {
  propertyID: ID!
  value: String!
}

input IntegerFilter {
  propertyID: ID!
  value: Integer!
}

input union Filter = StringFilter | IntegerFilter
```
This could probably also be considered a further generalization of the "Single user-chosen field; value is a literal" listed earlier.

Note: I think this approach can get complicated with coercion and nested types so I'm not advocating for it but if we're trying to catalog possible solutions, it's probably worth adding to the list.